kong/tools/queue.lua

```diff
@@ -0,0 +1,400 @@
+
+  local name = queue_conf.name
+
+  local queue = queues[name]
+  if queue then
+    queue:log_debug("queue exists")
+    -- We always use the latest configuration that we have seen for a queue and handler.
+    queue.handler_conf = handler_conf
+    return queue
+  end
+
+  queue = {
+    -- Queue parameters from the enqueue call
+    name = name,
+    handler = handler,
+    handler_conf = handler_conf,
+
+    -- semaphore to count the number of items on the queue and synchronize between enqueue and handler
+    semaphore = semaphore.new(),
+
+    -- `bytes_queued` holds the number of bytes on the queue.  It will be used only if max_bytes is set.
+    bytes_queued = 0,
+    -- `entries` holds the actual queue items.
+    entries = {},
+    -- Pointers into the table that holds the actual queued entries.  `front` points to the oldest, `back` points to
+    -- the newest entry.
+    front = 1,
+    back = 1,
+  }
+  for option, value in pairs(queue_conf) do
+    queue[option] = value
+  end
+
+  queue = setmetatable(queue, Queue_mt)
+
+  kong.timer:named_at("queue " .. name, 0, function(_, q)
+    while q:count() > 0 do
+      q:log_debug("processing queue")
+      q:process_once()
+    end
+    q:log_debug("done processing queue")
+    queues[name] = nil
+  end, queue)
+
+  queues[name] = queue
+
+  queue:log_debug("queue created")
+
+  return queue
+end
+
+
+-------------------------------------------------------------------------------
+-- Log a message that includes the name of the queue for identification purposes
+-- @param self Queue
+-- @param level: log level
+-- @param formatstring: format string, will get the queue name and ": " prepended
+-- @param ...: formatter arguments
+function Queue:log(handler, formatstring, ...)
+  local message = string.format("queue %s: %s", self.name, formatstring)
+  if select('#', ...) > 0 then
+    return handler(string.format(message, unpack({...})))
+  else
+    return handler(message)
+  end
+end
+
+function Queue:log_debug(...) self:log(kong.log.debug, ...) end
+function Queue:log_info(...) self:log(kong.log.info, ...) end
+function Queue:log_warn(...) self:log(kong.log.warn, ...) end
+function Queue:log_err(...) self:log(kong.log.err, ...) end
+
+
+function Queue:count()
+  return self.back - self.front
+end
+
+
+-- Delete the frontmost entry from the queue and adjust the current utilization variables.
+function Queue:delete_frontmost_entry()
+  if self.max_bytes then
+    -- If max_bytes is set, reduce the currently queued byte count by the
+    self.bytes_queued = self.bytes_queued - #self.entries[self.front]
+  end
+  self.entries[self.front] = nil
+  self.front = self.front + 1
+  if self.front == self.back then
+    self.front = 1
+    self.back = 1
+  end
+end
+
+
+-- Drop the oldest entry, adjusting the semaphore value in the process.  This is
+-- called when the queue runs out of space and needs to make space.
+function Queue:drop_oldest_entry()
+  assert(self.semaphore:count())
```

Comment:
Maybe you want to do this?

```suggestion
  assert(self.semaphore:count() > 0)
```

The number `0` is a truthy value in Lua.

Reply:
The count could be negative, which would also be a reason to fail the assertion.